File: Materials/Image/Android/app/src/main/java/com/dbooker/livreopengl/Ground.java
package com.dbooker.livreopengl;

import com.dbooker.livreopengl.glmatrix.mat4;
import com.dbooker.livreopengl.mesh.Mesh;
import com.dbooker.livreopengl.mesh.MeshModuleLoading;
import com.dbooker.livreopengl.mesh.MeshModuleDrawing;
import com.dbooker.livreopengl.material.VBOset;
import com.dbooker.livreopengl.material.TextureMaterial;
import com.dbooker.livreopengl.material.Texture2D;


// Définition de la classe Ground
public class Ground
{
    private Texture2D m_Texture;
    private TextureMaterial m_Material;
    private Mesh m_Mesh;
    private VBOset m_VBOset;

    /**
     * Constructeur
     */
    public Ground() throws Exception
    {
        // créer le matériau du terrain
        m_Texture = new Texture2D("data/models/Terrain/TerrainAt.png");
        m_Material = new TextureMaterial(m_Texture);

        // charger le maillage
        m_Mesh = new Mesh("Ground");
        MeshModuleLoading loader = new MeshModuleLoading(m_Mesh);
        loader.loadObjFile("data/models/Terrain/Terrain.obj", null, 1.0f);

        // créer le VBOset
        MeshModuleDrawing renderer = new MeshModuleDrawing(m_Mesh);
        m_VBOset = renderer.createStripVBOset(m_Material, true);
    }


    /**
     * dessine l'objet sur l'écran
     * @param mat4Projection : matrice de projection
     * @param mat4ModelView : matrice de vue
     */
    public void onDraw(mat4 mat4Projection, mat4 mat4ModelView)
    {
        m_VBOset.onDraw(mat4Projection, mat4ModelView);
    }


    /**
     * Cette méthode supprime les ressources allouées
     */
    public void destroy()
    {
        m_Mesh.destroy();
        m_Material.destroy();
        m_Texture.destroy();
        m_VBOset.destroy();
    }
}

File: Materials/Image/WebGL/Ground.js
// Définition de la classe Ground

// superclasses et classes nécessaires
Requires("libs/Material/TextureMaterial");
Requires("libs/Mesh/Mesh.js");
Requires("libs/Mesh/MeshModuleLoading.js");
Requires("libs/Mesh/MeshModuleDrawing.js");


class Ground
{
    /**
     * Constructeur
     */
    constructor()
    {
        // créer le matériau du terrain
        this.m_Texture = new Texture2D("data/models/Terrain/TerrainAt.png");
        this.m_Material = new TextureMaterial(this.m_Texture);

        // charger le maillage
        this.m_Mesh = new Mesh("Ground");
        let loader = new MeshModuleLoading(this.m_Mesh);
        loader.loadObjFile("data/models/Terrain/Terrain.obj", null, 1.0, this, this.onGroundLoaded, this.m_Material);
        this.m_VBOset = null;
    }


    onGroundLoaded(that, material)
    {
        // créer le VBOset
        let renderer = new MeshModuleDrawing(that.m_Mesh);
        that.m_VBOset = renderer.createStripVBOset(material, true);
    }


    /**
     * dessine l'objet sur l'écran
     * @param mat4Projection : matrice de projection
     * @param mat4ModelView : matrice de vue
     */
    onDraw(mat4Projection, mat4ModelView)
    {
        // dessiner le terrain s'il est prêt
        if (this.m_VBOset != null) {
            this.m_VBOset.onDraw(mat4Projection, mat4ModelView);
        }
    }


    /**
     * Cette méthode supprime les ressources allouées
     */
    destroy()
    {
        this.m_Mesh.destroy();
        this.m_Material.destroy();
        this.m_Texture.destroy();
        this.m_VBOset.destroy();
    }
}
